keps/sig-cluster-lifecycle/draft-20180316-kubeadm-phases-beta.md

```diff
@@ -0,0 +1,243 @@
+  with the continuous changes of the init workflow.
+* To provide a centralized point for controlling the proliferation of flags in
+  kubeadm.
+* To avoid inconsistencies between init and phases (both at UX and implementation
+  logic level)
+
+### Non-goals
+
+* This proposal doesn't include any changes/improvement to the init workflow
+  internal logic.
+* Nothing in this proposal should prevent practices that exist today.
+
+### Challenges and Open Questions
+
+* How to reduce the effort for maintaining `kubeadm phase` feature up to date
+  with the continuous changes of the init workflow?
+* How to provide an intuitive user interface for both phases and sub-phases
+  (e.g  `kubeadm alpha phase certs` and `kubeadm alpha phase certs apiserver`)?
+* Should we preserve all the second/third level nested phases implemented today
+  (e.g. certs CA, certs api-server, certs sa etc.), or should we allow only second
+  level phases (e.g certs), or should we maintain & simplify second/third level phases?
+* Should we preserve or deprecate the (few) phases implemented today that are not
+  part of the `kubeadm init` workflow? If preserve, in which form? as a new first
+  level commands?
+* While `kubeadm init` executes some phases conditionally (e.g. self-hosting only
+  if the corresponding feature-gate enable), the current implementation of phases
+  executes them unconditionally. Should we preserve this behavior?
+* How to preserve the possibility to automatically generate kubeadm documentation?
+* How to address/enable the possibility to generate machine readable output from
+  kubeadm commands?
+
+## Proposal
+
+### User Stories
+
+#### Invoke a specific task/phase of the kubeadm init workflow
+
+As a kubernetes administrator, I want to run only a specific task/phase of the
+`kubeadm init` workflow.
+
+### Implementation Details
+
+In order to achieve the goals of this proposal the approach suggested is to customize
+the `kubeadm init` command transforming it's usage from:
+
+```bash
+Usage:
+  kubeadm init [flags]
+```
+
+to
+
+```bash
+Usage:
+  kubeadm init [phases] [flags]
+```
+
+[phases] will be a new, optional, argument that allows user to select which part/phase
+of the `kubeadm init`  logic to execute.
+
+As a consequence:
+
+* To invoke the beta (GA) version of  `kubeadm alpha phase certs` the user will be required
+  to execute:
+
+```bash
+kubeadm init certs [flags]
+```
+
+* To invoke the beta (GA) version of  `kubeadm alpha phase certs apiserver` the user
+  will be required to execute:
+
+```bash
+kubeadm init certs/apiserver [flags]
+```
+
+NB. By executing `kubeadm init`  without [phases], all the task/phases which are part
+of the init workflow will be executed; this will guarantee consistency with the
+current behavior.
+
+#### Expected benefits
+
+The proposed implementation approach is expected to give following advantages:
+
+* `kubeadm phase` "by design" will be aligned to any change to `kubeadm init`
```

Comment:
Why two commands? This is very confusing.

Reply:
Good point. `kubeadm phases` should be replaced with `kubeadm init [phases]`